pkg/apis/serving/v1beta1/transformer_custom.go

```diff
@@ -62,6 +62,10 @@ func (c *CustomTransformer) GetStorageUri() *string {
 	return nil
 }
 
+func (c *CustomTransformer) GetStorageSpec() *StorageSpec {
```

Comment:
So for custom transformer the storage spec is not supported? Currently it is actually supported via `STORAGE_URI` env variable

Reply:
@yuzisun I'm not sure... TBH I had to look again at how custom predictors/transformers are handled. It seems to me that some of the things in `PredictorExtensionSpec` kind of belong up a level i.e. would be applicable to _any_ predictor including custom ones (at least `StorageURI`, `StorageSpec`, `ProtocolVersion`).
Similarly wouldn't it make sense to have StorageURI/Spec as strongly typed fields in the transformer spec too?

Semi-related, is the intention to keep a separate `Container` embedded in the `PredictorExtensionSpec` in addition to the `PodSpec`? Was wondering what the interplay might be with the `ServingRuntime` case where the SR might have multiple containers.

Reply:
For the `custom` framework, because it doesn't use the `PredictorExtensionSpec`, the new storageSpec will not be reflected over here. The new `model` spec for predictor should have `PredictorExtensionSpec` for all frameworks, so we need to have the same `model` spec for explainer and transformer so that their API spec won't get diverged based on the framework.

Reply:
@njhill The original design thinking is that `predictor` maps to the `pod` level fields while `model` maps to `container` level fields, that's why we embed a single `Container` on `PredictorExtensionSpec` and does not allow user set container fields on `PodSpec` if model spec presents. We want user to only be able to customize model container and sidecar should be always managed by KServe, I guess there might be still use case when someone wants to define SR with their sidecar implementation.

Regarding `StorageURI`/spec, we originally had the debate of making this field strongly typed for custom predictor/transformer cases but realized the spec looks a bit weird if we level up the field.
```yaml
spec:
   predictor:
     storageUri: s3://xxx
     sklearn:
```
vs
```yaml
spec:
    predictor:
       sklearn:
           storageUri: s3://xxxx
```
As we introduced model spec, this actually look a bit better, still feel storage should be part of the model spec but maybe this is not a big deal as we do name it generically as `storage`.
```yaml
spec:
    predictor:
       storageUri: s3://xxxx
       model:
          modelFormat:
               name: sklearn
```

Reply:
OK thanks @yuzisun

I guess in my mind it makes sense details about the persisted model (format, storage location, etc.) are kind of separate to properties related to how the model is deployed (i.e. the kind of predictor, framework-specific deployment config params, pod/container customizations), and so make sense for these things to sit alongside each other. I guess this will become cleaner when we move away from the strongly typed frameworks in favour of the serving runtimes.

> I guess there might be still use case as you said when someone wants to define SR with their own sidecar implementation.

Yes it could be the case that the overall runtime implementation takes the form of multiple containers. In particular if someone wanted to integrate some existing model server that we don't currently support, they could implement their own thin adapter container to be used in conjunction with it.